flytekit/extras/tasks/shell.py

```diff
@@ -97,6 +98,35 @@ def interpolate(
 T = typing.TypeVar("T")
 
 
+class MappingRevisor:
+   def __init__(self, mapping):
+       self.mapping = mapping
+       self.changes = {}
+       self.new_keys = []
+
+   def __getitem__(self, key):
+       return self.mapping[key]
+
+   def __setitem__(self, key, value):
+       if key not in self.mapping:
+           self.new_keys.append(key)
+       if key not in self.changes and key in self.mapping:
+          self.changes[key] = self.mapping[key]
+       self.mapping[key] = value
+
+   def reset(self):
+       for k, v in self.changes.items():
+           self.mapping[k] = v
+       for k in self.new_keys:
+           self.mapping.pop(k)
+
+   def __enter__(self):
+       return self
+
+   def __exit__(self, *args, **kwargs):
+       self.reset()
```

Comment:
if you think reset is idempotent enough (I think it is?) I guess the __del__ wouldn't be bad, and buttresses against misuse?

Reply:
ya, that's what I was thinking. `reset()` should be idempotent. Will add del now